Allow referencing bundle resources by name

andrewnester opened the pull request:
## Changes
Now we can define variables with values which reference different Databricks resources by name.
When references like this, DABs automatically looks up the resource by this name and replaces the reference with ID of the resource referenced. Thus when the variable is used in the configuration it will contain the correct resolved ID of resource.

The resolvers are code generated and thus DABs support referencing all resources which has `GetByName`-like methods in Go SDK.

### Example

```
variables:
  my_cluster_id:
    description: An existing cluster.
    lookup: 
      cluster: "12.2 shared"

resources:
  jobs:
    my_job:
      name: "My Job"
      tasks:
        - task_key: TestTask
          existing_cluster_id: ${var.my_cluster_id}

targets:
  dev:
    variables:
      my_cluster_id:
        lookup: 
           cluster: "dev-cluster"
```

## Tests
Added unit test + manual testing

pietern commented on .codegen/resolvers.go.tmpl:
```diff
@@ -0,0 +1,37 @@
+func Resolvers() map[string](ResolverFunc) {
+    resolvers := make(map[string](ResolverFunc), 0)
+    {{range .Services -}}
+    {{- if not .IsAccounts -}}
```
Conversely, is there an `IsWorkspace` that we can use?

andrewnester replied:
@pietern unfortunately not

pietern commented on bundle/config/mutator/resolve_resource_references_test.go:
```diff
@@ -0,0 +1,202 @@
+	"github.com/stretchr/testify/require"
+)
+
+type MockClusterService struct{}
```
How did you generate these mocks? I think having a structured way to (auto-) generate them would be good to have, because as-is, any time a method is added to the upstream service, it would need to be manually reflected here.

That, and it would be great to use client mocking in more tests.

andrewnester replied:
@pietern just with quick fix suggestion in IDE.

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+var separator string = ":"
+
+type resolveResourceReferences struct {
+	resolvers map[string](resolvers.ResolverFunc)
```
Parenthesis not required (I even expected gofmt to remove them).

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/databricks/cli/libs/log"
+)
+
+var separator string = ":"
```
Can be a const.

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+		}
+
+		lookup := v.Lookup
+		parts := strings.Split(lookup, separator)
```
Can use `strings.Cut` to force 2 output chunks (also allows use of `:` in the name).

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+		resource, name := parts[0], parts[1]
+		resolver, ok := m.resolvers[resource]
+		if !ok {
+			return fmt.Errorf("unable to resolve resource reference %s, no resovler for %s", lookup, resource)
```
Typo

pietern replied:
For future reference, it would be super if we can use `maps.Keys(m.resolvers)` into some kind of "did you mean XYZ" error message in case of typos.

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func (m *resolveResourceReferences) Apply(ctx context.Context, b *bundle.Bundle) error {
+	for k, v := range b.Config.Variables {
```
I think we can/should parallelize these lookups. You could use `sync.WaitGroup` (also see `libs/sync` for parallelization). They are inherently independent. Can also be a follow up.

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,61 @@
+
+		id, err := resolver(ctx, b, name)
+		if err != nil {
+			return fmt.Errorf("failed to resolve resource reference %s, err: %w", lookup, err)
```
```suggestion
			return fmt.Errorf("failed to resolve %s reference %s, err: %w", resource, lookup, err)
```

pietern commented on bundle/resolvers/resolvers.go:
```diff
@@ -0,0 +1,253 @@
+		}
+
+		return fmt.Sprint(entity.ObjectId), nil
+	}
```
Not all of these (majority?) are not relevant in the context of DABs. I would prefer an allow-list to ensure we only expose lookups that make sense. For ex, clusters, instance pools, warehouses, are all great. But the workspace lookup is (I believe) a lookup by path. All the UC related ones as well, they are already named so don't need a lookup. The tokens and token management ones are not relevant.

pietern reviewed (commented):
Nice work!

Is it possible to write:
```yaml
variables:
  foo:
    lookup: cluster_policy:${var.cluster_policy_name}
```

And it still works? It might, if we allow interpolation on `variables` before running the lookup mutator.

andrewnester commented:
> And it still works? It might, if we allow interpolation on variables before running the lookup mutator.

@pietern At the moment it doesn't work. The reason is that lookups are also variables and to make variables inside lookup work we will either need to do interpolation twice (before and after lookup) or incorporate the lookup into interpolation.
Overall it makes the logic more complex and we should better allow override lookups per targets

shreyas-goenka reviewed (commented):
This is pretty cool!

lennartkats-db commented on .codegen/resolvers.go.tmpl:
```diff
@@ -0,0 +1,54 @@
+import (
+	"context"
+    "fmt"
+	"github.com/databricks/cli/bundle"
```
Is this file mixing tabs and spaces?

lennartkats-db commented on .codegen/resolvers.go.tmpl:
```diff
@@ -0,0 +1,54 @@
+        "metastores"
+        "pipelines"
+        "queries"
+        "warehouses"
```
These are great! But could you also support service principals and webhooks as seen in https://github.com/databricks/universe/blob/master/serverless-smoke-tests/databricks.yml?

lennartkats-db replied:
@andrewnester could you still take a look at this? Are those resource types supported or not?

andrewnester replied:
@lennartkats-db added service principals but it seems like webhook notifications does not have an API (at least the one defined in OpenAPI) hence no support yet.

lennartkats-db replied:
I see. It's disappointing we can't support webhooks yet, but we should indeed do that with a proper public API. Let's push on getting that API support. I tagged you on a related thread.

pietern commented on bundle/config/target.go:
```diff
@@ -30,10 +30,10 @@ type Target struct {
 	Resources *Resources `json:"resources,omitempty"`
 
-	// Override default values for defined variables
+	// Override default values or lookup naem for defined variables
```
Typo

pietern commented on bundle/config/target.go:
```diff
@@ -30,10 +30,10 @@ type Target struct {
 	// Does not permit defining new variables or redefining existing ones
 	// in the scope of an target
-	Variables map[string]string `json:"variables,omitempty"`
+	Variables map[string]interface{} `json:"variables,omitempty"`
```
Use `any` instead of `interface{}`

pietern commented on bundle/tests/variables/env_overrides/databricks.yml:
```diff
@@ -6,6 +6,10 @@ variables:
 
+  d:
+    description: variable with lookup
+    lookup: "clusters: some-cluster"
```
Using this for a bit, I realized the prefixed string is odd given that we have a structured for everything else. It also doesn't help discovery/documentation, as we can't autocomplete this.

I suspect this is a remnant of earlier iterations where we lumped this into the `default` key with a lookup string. Then we later figured to break it out into a separate key but didn't revisit whether it should be a string or not.

However, if we break out the lookup type into an object with a key for every resolver, we get discovery and documentation for free.

andrewnester replied:
Added structured lookup field

pietern reviewed (commented):
Tried this PR out with a sample bundle and ran into the following:

* Singular vs plural; I tried out `instance-pool` first and got `no resolvers for instance-pool`, but had to specify the plural name. This is doing a lookup of a single instance so I think we should use singular in the prefix.
* I prefixed the name of the pool with a space, like this: `instance-pools: pieter's test pool` and got an error saying it doesn't exist (with a leading space). This can (should?) be trimmed. If we break it out into a separate field, this ambiguity would be gone (default YAML would trim leading spaces, but you could force it to use leading spaces by quoting a string).
* Using a parameterized name doesn't work yet, and it looks like it should work:
```yaml
variables:
  instance_pool_name:
    default: pieter's test pool

  instance_pool_id:
    lookup: "instance-pools:${var.instance_pool_name}"
```

andrewnester commented:
@pietern addressed points 1 and 2 by using typed structure for Lookup field. As to using variables in lookups it is currently not supported, it would require multiple interpolation passes. Instead we emit an error when variable reference is used in lookup field

pietern commented on bundle/config/mutator/resolve_resource_references.go:
```diff
@@ -0,0 +1,48 @@
+	errs, errCtx := errgroup.WithContext(ctx)
+
+	for k := range b.Config.Variables {
+		v := b.Config.Variables[k]
```
v can be nil as well (can happen if the variable is defined but doesn't have a default or a description)

pietern commented on bundle/config/mutator/resolve_resource_references_test.go:
```diff
@@ -0,0 +1,197 @@
+// UpdatePermissions implements compute.ClustersService.
+func (MockClusterService) UpdatePermissions(ctx context.Context, request compute.ClusterPermissionsRequest) (*compute.ClusterPermissions, error) {
+	panic("unimplemented")
+}
```
TBD: migrate this to the new SDK mocks when available in this repo

andrewnester replied:
@pietern added to do list :)

pietern commented on bundle/config/variable/lookup.go:
```diff
@@ -0,0 +1,277 @@
+type Lookup struct {
+	Alert string `json:"alert,omitempty"`
+
+	ClusterPolicy string `json:"cluster-policy,omitempty"`
```
We should use lower case for these field names to stay consistent with the rest of the config.

pietern commented on bundle/config/root.go:
```diff
@@ -201,13 +201,25 @@ func (r *Root) MergeTargetOverrides(target *Target) error {
+				// we  allow overrides of the default value for a variable
+				defaultVal := sv
+				rootVariable.Default = &defaultVal
+			} else if vv, ok := v.(map[string]interface{}); ok {
```
interface{} -> any

pietern commented on bundle/config/root.go:
```diff
@@ -201,13 +201,25 @@ func (r *Root) MergeTargetOverrides(target *Target) error {
+				if !ok {
+					return fmt.Errorf("variable %s is incorrectly defined lookup override, no 'lookup' key defined", k)
+				}
+				rootVariable.Lookup = variable.LookupFromMap(lookup.(map[string]interface{}))
```
Same

pietern commented on .codegen/lookup.go.tmpl:
```diff
@@ -0,0 +1,122 @@
+type Lookup struct {
+	{{range .Services -}}
+	{{- if in $allowlist .KebabName -}}
+	{{.Singular.CamelName}} string
```
This was OK with `PascalName`, but the JSON tag needs `SnakeName` (don't know if that's a thing)

pietern reviewed (approved):
Unblock
